feat(observability): Add the Observability Plugin as part of kork-actuator

mrusan opened the pull request:
# feat(observability): Donate Armory Observability Plugin to kork-actuator

## Summary

Migrates the [Armory Observability Plugin](https://blog.spinnaker.io/announcing-the-new-spinnaker-observability-plugin-d7fbb17e1e07) into `kork-actuator` as first-party Spinnaker functionality. This PR supersedes [spinnaker/kork#1227](https://github.com/spinnaker/kork/pull/1227).

## Changes

| Area | Details |
|------|---------|
| Migration | Observability configuration, registry suppliers (Prometheus, Datadog, New Relic), scrape endpoint, recommended metric filters, and `TagsService` added under `kork-actuator` |
| Auto-configuration | `EnvironmentPostProcessor` entries in `spring.factories`, `ObservabilityConfiguration` in `AutoConfiguration.imports` |
| Naming | Removed Armory-branded identifiers; endpoint standardized to `/actuator/prometheus` |
| Exporter guardrails | `DatadogMetricsPostProcessor` and `PrometheusMetricsPostProcessor` no longer override Boot exporter flags when `observability.enabled=false` or when operator has set `management.metrics.export.*` explicitly |
| New Relic guard | Fail-fast at startup when `override-primary-registry=false` + New Relic enabled (unsupported combination) |
| Test layout | Tests under standard `src/test`; no custom sourceSet wiring |
| Test stability | Prometheus integration test uses pinned `prom/prometheus:v2.53.3`, bounded 30s poll, mockserver 5.15.0 (JUnit 5), rest-assured 5.3.2, testcontainers from BOM |
| Documentation | `kork-actuator/src/main/docs/observability-plugin.md` covers configuration, ownership model, provider notes, and troubleshooting |

### Files Added (observability)

* `ObservabilityConfiguration.java` — main auto-configuration and composite registry wiring
* `DatadogMetricsPostProcessor.java` / `PrometheusMetricsPostProcessor.java` — environment post-processors with exporter guardrails
* `PrometheusRegistrySupplier.java` / `DataDogRegistrySupplier.java` / `NewRelicRegistrySupplier.java` — registry suppliers
* `PrometheusScrapeEndpoint.java` — WebEndpoint serving `/actuator/prometheus`
* `MutatedPrometheusMeterRegistry.java` — customized Prometheus registry
* `TagsService.java` / `MeterFilterService.java` / `RecommendedFilters.java` — tags and filtering
* `ObservabilityConfigurationProperties.java` + model classes — config binding

### Files Modified

* `kork-actuator.gradle` — added observability dependencies (micrometer registries, lombok, test libs)
* `spring.factories` / `AutoConfiguration.imports` — registered post-processors and auto-configuration

## Backward Compatibility

Observability is opt-in via `observability.enabled=true`. Services without this property are unaffected — the exporter post-processors exit early and `ObservabilityConfiguration` does not activate.

## Test Coverage

* Post-processor behavior (enabled/disabled/explicit-override scenarios)
* Config context tests (bean wiring under various property combinations)
* API key masking, tags service, meter filter service
* Registry supplier invocation and endpoint conditionals
* Prometheus integration test with Testcontainers + MockServer

christosarvanitis committed:
feat(observability): Migrate Observability plugin to kork

danielGz committed:
refactor(observability): Refactor observability plugin code to be autonomous

danielGz committed:
refactor(observability): Modifications for observability functionality autonomy

danielGz committed:
refactor(observability): Removed unnecessary file

danielGz committed:
test(observability): Added tests

danielGz committed:
docs(observability): Fix copyright headers, tie datadog config from actuator to observability config

mrusan committed:
chore(kork-actuator): Wire lombok and include src/main/test; observability plugin migration follow-up

mrusan committed:
test(observability): Extend test coverage for config options

mrusan committed:
fix(observability): Fix typo

mrusan committed:
fix(observability): Match post-processor name with actual config

mrusan committed:
test(observability): Extend test coverage for post-processor

mrusan committed:
fix(observability): Fix transposition [i-t] in ObservabilityConfigurationProperties

mrusan committed:
refactor(observability): Use correct header

mrusan committed:
fix(observability): Build properties once - at construction time

mrusan committed:
docs(observability): Port and update documentation from the Armory Observability plugin

mrusan committed:
fix(observability): Invert conditional logic to be correct

mrusan committed:
test(observability): Test added for tag conditional logic

mrusan committed:
refactor(observability): Remove automatically handled nested properties

mrusan committed:
refactor(observability): Disable Spring Boot exporters to avoid duplicate registries

mrusan committed:
refactor(observability): Prevent API key exposure via actuator endpoints

mrusan committed:
refactor(observability): Ensure JUnit 4 tests run on JUnit Platform

mrusan committed:
refactor(observability): Match imports and interface with New Relic Telemetry 0.5.0 package names

jasonmcintosh commented on kork/kork-actuator/src/main/docs/observability-plugin.md:
````diff
@@ -0,0 +1,225 @@
+            - jvm.gc
+```
+
+- **Scrape endpoint**: `/actuator/aop-prometheus`
````
Question:  IS it time to change this?  specifically the dash in the actuator?

christosarvanitis commented on kork/kork-actuator/src/main/java/com/netflix/spinnaker/kork/actuator/observability/model/ArmoryEnvironmentMetadata.java:
```diff
@@ -0,0 +1,27 @@
+
+@Data
+@Builder
+public class ArmoryEnvironmentMetadata {
```
The extra metadata is not needed for OSS as this was adding tags to keep track of the OSS metadata. Consider removing this or repurposing (although extra tags mean extra load for Observability platforms)

kirangodishala commented on kork/kork-actuator/src/main/java/com/netflix/spinnaker/kork/actuator/observability/prometheus/PrometheusScrapeEndpoint.java:
```diff
@@ -0,0 +1,72 @@
+ * https://github.com/spring-projects/spring-boot/blob/cd1baf18fe9ec71c11d7d131d6f1a417ec0c00e2/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/export/prometheus/PrometheusScrapeEndpoint.java
+ */
+// If you use WebEndpoint instead of Endpoint, the plugin throws class def not found error with PF4j
+// ¯\_(ツ)_/¯
```
This may not be valid anymore. It can be a WebEndpoint now.

mrusan committed:
refactor(observability): rename endpoint to /actuator/prometheus, use WebEndpoint

- Change endpoint id from 'aop-prometheus' to 'prometheus'
- Switch from @Endpoint to @WebEndpoint (no longer PF4J plugin)
- Fix duplicate TextFormat.write004() call that wrote nothing
- Update javadoc to reflect current implementation

mrusan committed:
docs(observability): add attribution and update for new naming

- Add attribution to Armory, Inc. as original developers
- Document breaking changes for migration from plugin
- Update endpoint references: aop-prometheus -> prometheus
- Update property: armory-recommended-filters-enabled -> recommended-filters-enabled

mrusan committed:
refactor(observability): remove Armory branding, rename classes and properties

- Rename ArmoryRecommendedFilters -> RecommendedFilters
- Rename ArmoryEnvironmentMetadata -> EnvironmentMetadata
- Rename ArmoryObservabilityCompositeRegistry -> ObservabilityCompositeRegistry
- Rename property armoryRecommendedFiltersEnabled -> recommendedFiltersEnabled
- Fix typo: observabilityConfigurationProperites -> observabilityConfigurationProperties
- Fix SpringApplication import in PostProcessor files
- Update test resources to use OSS-standard build properties
- Remove io/armory/plugin test resource paths, use observability/
- Use isEmpty() instead of size() == 0
- Update LIB_NAME from 'aop' to 'kork-observability'

mrusan committed:
refactor(observability): clean up MutatedPrometheusMeterRegistry

- Add detailed fork rationale explaining why this class exists
- Remove unused tagValues() method and all variable assignments
- Remove stale commented-out code
- Add TODO to evaluate migrating to upstream when fixes are merged

mrusan committed:
fix(observability): cache supplier results to prevent duplicate registries

Suppliers were invoked twice in the constructor - once for building
the registries list and again for building the config map. This caused
duplicate registry creation and, for NewRelic, duplicate background
threads.

mrusan committed:
fix(observability): add conditionals to prometheus beans

Only create CollectorRegistry, PrometheusRegistrySupplier, and
PrometheusScrapeEndpoint when prometheus is enabled. Also prevent
endpoint collision with Spring Boot's prometheus endpoint when
override-primary-registry is false.

mrusan committed:
test(observability): add tests for supplier invocation and endpoint conditionals

Verify each supplier is called exactly once to prevent duplicate
registries. Add tests for prometheus endpoint conditional behavior.

mrusan committed:
docs(observability): fix breaking change notice with actual before/after values

The breaking change section had identical before/after values.
Updated with the actual changes from the plugin migration.

mrusan committed:
test(observability): add TagsService and API key masking tests

Verify default tags behavior, immutability, and that API keys are
not exposed when config objects are serialized to JSON.

mrusan committed:
fix(observability): remove unused parameter from private constructor

mrusan committed:
fix(observability): use anonymous class for VersionResolver in tests

VersionResolver extends Ordered, making it a non-functional interface
with two abstract methods. Replace lambda expressions with anonymous
inner classes to fix compilation.

mrusan committed:
fix(observability): use nested config for prometheus endpoint conditions

@ConditionalOnProperty is not repeatable. Move PrometheusScrapeEndpoint
to a nested @Configuration class to combine multiple property conditions.
Update test to include the nested configuration class.

mrusan committed:
fix(observability): guard exporter overrides and validate unsupported New Relic mode

mrusan committed:
fix(observability): bound scrape polling and hide default tag values in logs

mrusan committed:
docs(observability): keep ownership docs while dropping branch review artifact

mrusan committed:
refactor(kork-actuator): migrate observability tests to standard src/test layout

mrusan committed:
fix(dependencies): drop stale slf4j 1.7.36 strict pin

mrusan committed:
fix(build): revert JVM heap to 2g to match main

mrusan committed:
fix(kork-actuator): drop pinned testcontainers version, use BOM 1.19.8

mrusan committed:
fix(kork-actuator): bump rest-assured to 5.3.2 for Jakarta compatibility

mrusan committed:
fix(kork-actuator): migrate mockserver to 5.15.0 with JUnit 5 extension

mrusan committed:
fix(observability): disable Boot exporters when observability is off

The micrometer registry JARs ship transitively to all Spinnaker services
via kork-runtime → kork-actuator. Without this change, Spring Boot's
auto-configuration detects the registry classes on the classpath and
activates exporters (Prometheus, Datadog) even when observability is not
enabled, breaking services that never had these JARs before.

The post-processors now explicitly set management.metrics.export.<provider>.enabled=false
when observability.enabled is not true, preserving pre-migration behavior.
Operator-provided management.metrics.export.* values still take precedence.

mrusan committed:
Merge branch 'main' into donate-observability-plugin

mrusan committed:
Merge branch 'main' into donate-observability-plugin

mrusan committed:
Merge branch 'main' into donate-observability-plugin

jasonmcintosh committed:
Merge branch 'main' into donate-observability-plugin

jasonmcintosh committed:
Merge branch 'main' into donate-observability-plugin

mrusan committed:
style(kork-actuator): Apply spotless formatting

jasonmcintosh committed:
Merge branch 'main' into donate-observability-plugin

jasonmcintosh committed:
Merge branch 'main' into donate-observability-plugin

jasonmcintosh committed:
Merge branch 'main' into donate-observability-plugin

jasonmcintosh committed:
Merge branch 'main' into donate-observability-plugin

jasonmcintosh committed:
Merge branch 'main' into donate-observability-plugin

jasonmcintosh committed:
Merge branch 'main' into donate-observability-plugin

jasonmcintosh committed:
Merge branch 'main' into donate-observability-plugin

mrusan commented:
Closing in favor of future OTEL adoption.
